### PR TITLE
fix link doc to envir

### LINKED ISSUE #3717: ProxySpace.linkDoc is broken
When I try to link a `ProxySpace` (or some other "EnvironmentRedirect") to the current document, `linkDoc` fails: switching to another document, the ProxySpace is still pushed.

On the other hand, the direct setter works:

`Document.current.envir = ProxySpace.new(s);`

The reason is in a duplicate push (once in `linkDoc` and once in `envir_`).

### PR DESCRIPTION
**Pull request by telephon**
This fixes #3717.

The basic assumption is that only one document can be in focus.

1. when a document comes to focus that has an envir bound to it, it should make that envir the `currentEnvironment`. Whatever envir that was current at that time should be saved.
2. when a document loses focus (either by closing or because one moves to another one), it should restore the envir that was currentEnvironment before.
3. when we set the envir of a document that has no focus, it should not do anything but store the envir for the moment it comes to focus.
4. when we set the envir of a document that has focus, the envir should by default become `currentEnvironment`, so that it is in a consistent state.

### CODE REVIEW

**Comment by LFSaw** [comment]
can you provide a test case please? ;)

**Comment by LFSaw** [comment]
sorry, I was too fast... there's a test class and I'll have a look at how to deal with it (following advice in `UnitTest` documentation and brian's unit testing overview 
https://gist.github.com/brianlheim/91222d487afa18582c287b0a722ae272 ).

**Comment by telephon** [comment]
yes, it doesn't adhere to the "only one assert per method" rule. The resulting code would be much harder to read, because one test relies on the state of the previous.

**Comment by telephon** [comment]
It would be good if you could test it also "by hand", maybe just with `ProxySpace(s).linkDoc`, and see if you are happy with the behavior.

**Comment by telephon** [comment]
@LFSaw I've improved the tests a bit.

**Comment by LFSaw** [comment]
thanks, will look into it warliest monday.

**Comment by LFSaw** [comment]
(( side note if you name your PRs/branches `topic/fix/linkDoc-03` instead of  `topic-fix-linkDoc-03`, it is recognised by many (e.g. my preferred) git gui tool as folders in a tree overview. this makes searching for topics much easier for externals. ))

File: testsuite/classlibrary/TestDocumentEnvir.sc [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+
+	}
+
+
+
+
+}
+
+
+TestEnvironmentDocument : UnitTest {
```
Comment (LFSaw) [inline testsuite/classlibrary/TestDocumentEnvir.sc]: Is this the correct Naming scheme?

Comment (telephon) [reply testsuite/classlibrary/TestDocumentEnvir.sc]: I supposed that this is reasonable. So far, we haven't been very systematic about it.

**Comment by telephon** [comment]
re side note: github desktop doesn't permit slashes for whatever reason

**Pull request merged**